docs/snagfactory_config.md
# Configuring Snagfactory

The Snagfactory configuration file is a yaml document with two main sections:

The "boards" section, which specifies a list of USB vid:pid pairs that
correspond to board USB recovery devices. Each USB device is associated with an
SoC model supported by snagrecover.

```yaml
boards:
  "0451:6165": "am625"
  "1f3a:efe8": "a64"
```

The "soc-models" section, which specifies a list of recovery and factory
flashing parameters for each SoC model. The '<soc_name>-firmware' section
specifies a snagrecover firmware configuration for this SoC. Please refer to
[the snagrecover docs](snagrecover.md#preparing-recovery-firmware) for more information. The
'<soc_name>-tasks' section specifies a list of factory flashing actions to be
performed in order.

```yaml
soc-models:
  am625-firmware:
    tiboot3:
          path: "tiboot3_evm.bin"
    tispl:
          path: "tispl_evm.bin"
    u-boot:
          path: "u-boot_evm.img"

  am625-tasks:
    - target-device: mmc0
      fb-buffer-addr: 0xd0000000

    ...
```

## Factory flashing tasks

Each SoC model in the Snagfactory configuration is associated with a list of
tasks that will be run in order. If one of the tasks fails, the ones that follow
it will not run.

The first element in the '<soc_name>-tasks' list is a dictionary of global
configuration variables, that are common to all tasks. It must contain the
following values:

```yaml
target-device: The device configured as the Fastboot flashing backend in U-Boot. Either 'mmc<num>' or 'nand'.
fb-buffer-addr: The size in bytes of the Fastboot buffer.
eraseblk-size: The size in bytes of an erase block for MTD targets
fb-buffer-size: (optional) The size in bytes of the Fastboot buffer. This can only be used to reduce the default U-Boot buffer size.
```

These variables may be updated during the factory flashing process, e.g.:

```yaml
- target-device: mmc0
  fb-buffer-addr: 0xd0000000

- task: flash
  args: ...

# change target device to NOR flash
- target-device: nor0

- task: flash
  args: ...
```

The following elements are standard Factory flashing tasks. They all have the
following format:

```yaml
task: task name
args: task parameters
```

The following sections describe all tasks supported by Snagfactory.

## run

Action:

Runs a list of Fastboot commands, in snagflash format. Please refer to
[the snagflash docs](snagflash.md) for more information.

Example:

```yaml
task: run
args:
  - "getvar:version"
  - "download:image.bin"
  - "oem_run:mmc rescan"
  - "flash:1:3"
```

## gpt

Action:

Writes a GPT partition table to the Fastboot backend device and
optionally flashes images to the newly created partitions. For each partition,
you must specify at least its name and size. The following additional parameters
can also be specified:

Unless otherwise specified, all values are in bytes.

```
image: path of an image to flash to the partition
image-offset: offset to which the image must be flashed inside the partition
start: offset of the partition from the start of the MMC device.
bootable: indicates if the "bootable" GPT flag is set for this partition. Defaults to False.
uuid: GPT UUID of this partition
type: GPT type UUID of this partition
```

Example:

```
task: gpt
args:
  - name: boot
    size: 3M
    bootable: True
  - name: rootfs
    size: 1000M
    image: "./rootfs.ext4"
```

## mtd-parts

Action:

Sets the "mtdparts" U-Boot environment variable, which describes a fixed
partition layout on an MTD device. Please note that this will be cleared when
the board is reset. Images can optionally be flashed to the newly created
partitions.For each partition, you must specify at least its name and size. The
following additional parameters can also be specified:

Unless otherwise specified, all values are in bytes.

```
image: path of an image to flash to the partition
image-offset: offset to which the image must be flashed inside the partition
start: offset of the partition from the start of the MTD device.
ro: indicates if the "read-only" flag is set for this partition. Defaults to False.
```

Example:

```
- task: mtd-parts
  args:
    - name: "ospi.tiboot3"
      size: 512k
    - name: "ospi.tispl"
      size: 2m
    - name: "ospi.u-boot"
      size: 4m
    - name: "ospi.env"
      size: 256k
    - name: "ospi.rootfs"
      image: "big_nand_image"
      size: 98048k
      start: 32m
```


## flash

Action:

Flashes a binary image to a partition. The partition name can be an index, the
name of a GPT partition on the backend device, or the name of an MTD device or
partition. The "hwpart <hwpart_num>" syntax can be used with the "part" parameter
to target a hardware partition on an MMC device.

**Note**: the "part: hwpart <part_num>" syntax can only be used with an eMMC backend
**Note**: "part: hwpart 0" targets the user area, "part: hwpart 1" targets the first eMMC boot partition, and so on...


Example:

```
task: flash
args:
  - part: rootfs
    image: "rootfs.ext4"
  - part: "hwpart 1"
    image: "boot.bin"
    image-offset: 0x200
```

## reset

Action:

Runs the U-Boot "reset" command on the target and recovers the board with
snagrecover afterwards. This can be useful for flashing actions which require a
reset to become effective.

**Note**: Some flashing tasks, such as "mtd-parts", affect the U-Boot
environment. Resetting the board will clear these changes.

Example:

```
task: reset
```

## prompt-operator

Action:

Pauses the task pipeline for this device and displays a message for the operator
to read. The task pipeline will be resumed once the operator clicks on the
"resume" button for the board. The "reset-before" flag can be added to run a
U-Boot reset command before the operator resumes the board and recover it afterwards.

Example:

```
task: prompt-operator
args:
  prompt: "please power-cycle the board"
  reset-before: True
```

## emmc-hwpart

/!\
**WARNING**: This task applies irreversible changes to your devices! Make sure to read the [hardware partitioning guide](hw_partitioning.md) carefully before using it!
/!\

Action:

Configures hardware partitions on an eMMC device. This is irreversible once
completed. Please read the [hardware partitioning guide](hw_partitioning.md)
carefully before using this task!

**Warning**: All start and size arguments are specified in **bytes** and not 512-byte blocks!

Example:
```
  euda:
    start: 0
    size: 0x258000
    wrrel: False
  gp1:
    size: 0x470000
    enh: true
    wrrel: True
  gp2:
    size: 0x470000
    enh: true
    wrrel: True

  skip-pwr-cycle: False
```



src/snagrecover/config.py
# This file is part of Snagboot
# Copyright (C) 2023 Bootlin
#
# Written by Romain Gantois <romain.gantois@bootlin.com> in 2023.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

import os
import yaml
from snagrecover.utils import (
	cli_error,
	usb_addr_to_path,
	get_family,
	access_error,
	resolve_soc_model,
)
import logging

logger = logging.getLogger("snagrecover")

default_usb_ids = {
	# default ROM code USB IDs
	"stm32mp": "0483:df11",
	"sama5": "03eb:6124",
	"sunxi": "1f3a:efe8",
	"am6x": {
		"am625": "0451:6165",
		"am62a7": "0451:6165",
		"am62d2": "0451:6165",
		"am62p": "0451:6165",
		"am6442": "0451:6165",
		"am654x": "0451:6165",
		"am623": "0451:6165",
		"am6411": "0451:6165",
		"am6412": "0451:6165",
		"am6421": "0451:6165",
		"am6422": "0451:6165",
		"am6441": "0451:6165",
		"am6548": "0451:6162",
		"am6546": "0451:6162",
	},
	"am62lx": "0451:6165",
	"zynqmp": "03fd:0050",
	"keembay": "8087:0b39",
	"imx": {
		"imx8qxp": "1fc9:012f",
		"imx8qm": "1fc9:0129",
		"imx8dxl": "1fc9:0147",
		"imx28": "15a2:004f",
		"imx815": "1fc9:013e",
		"imx865": "1fc9:0146",
		"imx91": "1fc9:0159",
		"imx93": "1fc9:014e",
		"imx7d": "15a2:0076",
		"imx7s": "15a2:0076",
		"imx6q": "15a2:0054",
		"imx6d": "15a2:0061",
		"imx6sl": "15a2:0063",
		"imx6sx": "15a2:0071",
		"imx6ul": "15a2:007d",
		"imx6ull": "15a2:0080",
		"imx6sll": "1fc9:0128",
		"imx7ulp": "1fc9:0126",
		"imxrt106x": "1fc9:0135",
		"imx8mm": "1fc9:0134",
		"imx8mq": "1fc9:012b",
		"imx53": "15a2:004e",
	},
	"bcm": {"bcm2711": "0a5c:2711", "bcm2712": "0a5c:2712"},
	"amlogic": "1b8e:c003",
	"rzn1": "045b:0239",
}

recovery_config = {}  # Global immutable config to be initialized with CLI args


def complete_fw_paths(fw_config: dict, this_file_path: str = None) -> None:
	paths_relative_to_conf = fw_config.pop("paths-relative-to", "CWD")
	if paths_relative_to_conf == "CWD":
		return
	elif paths_relative_to_conf == "THIS_FILE":
		if this_file_path is None:
			cli_error(
				"'paths-relative-to: THIS_FILE' is only valid when passing a firmware configuration by path!"
			)

		path_relative_to = os.path.dirname(this_file_path)
	else:
		path_relative_to = paths_relative_to_conf

	for binary in fw_config.keys():
		if "path" in fw_config[binary]:
			fw_config[binary]["path"] = os.path.join(
				path_relative_to, fw_config[binary]["path"]
			)


def init_config(args: list):
	# this is the only time that config.recovery_config should be modified!
	# get soc model
	soc_model = resolve_soc_model(args.soc)
	recovery_config.update({"soc_model": soc_model})
	soc_family = get_family(soc_model)
	recovery_config.update({"soc_family": soc_family})

	if args.rom_usb is not None and args.usb_path is None:
		args.usb_path = args.rom_usb

	if soc_family != "am335x":
		if args.usb_path is None:
			usb_ids = default_usb_ids[soc_family]
			if isinstance(usb_ids, dict):
				usb_ids = usb_ids[soc_model]

			recovery_config["usb_path"] = usb_addr_to_path(usb_ids)

			if recovery_config["usb_path"] is None:
				access_error("USB", usb_ids)

		else:
			recovery_config["usb_path"] = usb_addr_to_path(args.usb_path)
			if recovery_config["usb_path"] is None:
				access_error("USB", args.usb_path)

	fw_configs = {}
	if args.firmware:
		for fw in args.firmware:
			if not isinstance(fw, dict):
				cli_error(
					"firmware config to CLI did not evaluate to Python3 dict: {fw}"
				)
			complete_fw_paths(fw)
			fw_configs = {**fw_configs, **fw}
		recovery_config["firmware"] = fw_configs
		if args.firmware_file:
			logger.warning(
				"You passed firmware configuration via files AND direct CLI arguments!"
			)
	if args.firmware_file:
		# get firmware configs
		for path in args.firmware_file:
			with open(path, "r") as file:
				fw_config_file = yaml.safe_load(file)
			if not isinstance(fw_config_file, dict):
				cli_error(
					f"firmware config passed to CLI did not evaluate to dict: {fw_config_file}"
				)
			complete_fw_paths(fw_config_file, path)
			fw_configs = {**fw_configs, **fw_config_file}
		recovery_config["firmware"] = fw_configs

	# store input arguments in config
	recovery_config["args"] = vars(args)
	logger.debug(f"recovery_config:{str(recovery_config)}")
